spectator-reg-metrics3/src/main/java/com/netflix/spectator/metrics3/MetricsRegistry.java

```diff
@@ -57,4 +54,42 @@ private String toMetricName(Id id) {
     final String name = toMetricName(id);
     return new MetricsTimer(clock(), id, impl.timer(name));
   }
+
+  @Override public <T extends Number> T gauge(Id id, T number) {
+    return gauge(id, number, (ToDoubleFunction<T>) value -> number.doubleValue());
+  }
+
+  @Override public <T> T gauge(Id id, T obj, ToDoubleFunction<T> f) {
+    final String name = toMetricName(id);
+
+    Gauge aggrGauge = this.impl.getGauges().get(name);
+
+    // Wasn't registered already
+    if (aggrGauge == null) {
+      aggrGauge = new MetricsGaugeAggr();
+      this.impl.register(name, aggrGauge);
```

Comment:
It looks like there is a race condition here, if two threads get before the register takes place, then both will get here and one will throw.